File: examples/ast/traverse_ast.cpp
//   Copyright (c) 2001-2017 Hartmut Kaiser
//
//   Distributed under the Boost Software License, Version 1.0. (See accompanying
//   file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

#include <phylanx/phylanx.hpp>

#include <iostream>

struct traverse_ast : phylanx::ast::static_visitor
{
    template <typename T>
    bool operator()(T const& val) const
    {
        std::cout << val << std::endl;
        return true;
    }
};

int main(int argc, char* argv[])
{
    phylanx::ast::primary_expr<double> p1{
        phylanx::ir::node_data<double>{3.14}};

    phylanx::ast::operand<double> op1{p1};
    phylanx::ast::expression<double> e1{op1};

    phylanx::ast::operation<double> u1{
        phylanx::ast::optoken::op_plus, std::move(op1)};
    e1.append(u1);

    std::list<phylanx::ast::operation<double>> ops = {u1, u1};
    e1.append(ops);

    phylanx::ast::traverse(e1, traverse_ast{});

    return 0;
}

File: tests/unit/python/ast/traverse_ast.py
#  Copyright (c) 2017 Hartmut Kaiser
#
#  Distributed under the Boost Software License, Version 1.0. (See accompanying
#  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

import phylanx

phylanx.PhylanxSession(1)

###############################################################################


class traverse_ast:
    def __init__(self):
        self.generated_string = ''
        pass

    def on_enter(self, ast):
        self.generated_string += str(ast) + '\n'
        return True

    # optional, can be ommitted
    def on_exit(self, ast):
        return True


def test_traverse(expr, expected):
    ast = phylanx.ast.generate_ast(expr)
    assert (len(ast) == 1)

    visitor = traverse_ast()
    phylanx.ast.traverse(ast[0], visitor)

    if visitor.generated_string != expected:
        print(visitor.generated_string, expected)
    assert (visitor.generated_string == expected)


###############################################################################
test_traverse('A + B', '(A$1$1 + B$1$5)\nA$1$1\nB$1$5\n+\n')
test_traverse('A + B - C',
              '(A$1$1 + B$1$5 - C$1$9)\nA$1$1\nB$1$5\n+\nC$1$9\n-\n')
